### PR TITLE
Fix wp-env docker-compose config with explicit users and passwords

### LINKED ISSUE #29752: Env: Access denied for user 'username_here'@'X.X.X.X' error running wp-env start.
## Description
When trying to start a new environment using `wp-env`, the CLI issues the following error: 
```
mysqlcheck: Got error: 1045: Access denied for user 'username_here'@'172.18.0.5' (using password: YES) when trying to connect
```

This issue is not specific to working locally using this repository, but exists wherever `@wordpress/env` is used.

## Step-by-step reproduction instructions
1. Destroy your local environment `npm run wp-env destroy`
2. Re-provision:  `npm run wp-env start`

Expect to see the errors. Should you not see the error, make sure your docker images are updated to the latest.

## Quick Debugging
- `username_here` was implemented in the WordPress docker image fairly recently (https://github.com/docker-library/wordpress/commit/8f5bcc1c980fce198bd191cfb7f51a3e76bb59cd#diff-7f28a0f62019f2d8b061638e8c3aea207404c1499f1c9241db3216fadd8f4530)
- the `wp-config.php` in the WordPress container matches the default config in the PR listed above.
- the `wp-config.php` includes the `getenv_docker` function that looks for a local `env_FILE` that does not exist in the WordPress container.
- When looking at the container using the docker desktop app, I see ``` No 'wp-config.php' found in /var/www/html, but 'WORDPRESS_...' variables supplied; copying '/usr/src/wordpress/wp-config-docker.php' (WORDPRESS_DB_NAME)```

## Environment Info

**Relevant NPM packages**
```
├─┬ @wordpress/env@3.0.2
│ ├── chalk@4.0.0 deduped
│ ├── copy-dir@1.3.0
│ ├── docker-compose@0.22.2
│ ├── extract-zip@1.6.7
│ ├── got@10.7.0
│ ├── inquirer@7.1.0 deduped
│ ├── js-yaml@3.13.1
│ ├── ora@4.0.2
│ ├── rimraf@3.0.2 deduped
│ ├── simple-git@2.35.0
│ ├── terminal-link@2.0.0
│ └── yargs@14.0.0
```

## WordPress information
trunk@https://github.com/WordPress/gutenberg/commit/647a41cbab8e44e5ecd79af381f54cec080a3758

### PR DESCRIPTION
**Pull request by youknowriad**
closes https://github.com/WordPress/gutenberg/issues/29752

There has been some update  to the WordPress default docker image with these two updates:

 - It doesn't create a database by default now.
 - It  doesn't use the default username/password for the database.

This PR fixes that  by assigning explicit values for these, the issue though is that all pre-existing wp-env usage is broken including third-party and old branches.

### CODE REVIEW

File: packages/env/lib/build-docker-compose-config.js [hunk 1]
```diff
@@ -176,9 +176,17 @@ module.exports = function buildDockerComposeConfig( config ) {
...
+				},
+			},
+			'tests-mysql': {
+				image: 'mariadb',
+				ports: [ '3306' ],
+				environment: {
+					MYSQL_ROOT_PASSWORD: 'password',
+					MYSQL_DATABASE: 'tests-wordpress',
 				},
-				volumes: [ 'mysql:/var/lib/mysql' ],
```
Comment (Clorith) [inline packages/env/lib/build-docker-compose-config.js:181]: The volume for the `mysql` container should be retained. Removing this means any restart of the environment loses any existing data.

The `clean` command exists to wipe databases on the dev environment when needed.

Comment (youknowriad) [reply packages/env/lib/build-docker-compose-config.js:181]: I think restarting the environment doesn't mean losing data, destroying the environment means that and for a dev environment, I think that's totally acceptable.

Comment (youknowriad) [reply packages/env/lib/build-docker-compose-config.js:181]: Regardless, I restored volumes to see if it solves the permission issues

File: packages/env/lib/build-docker-compose-config.js [hunk 2]
```diff
@@ -176,26 +176,41 @@ module.exports = function buildDockerComposeConfig( config ) {
 				image: 'mariadb',
 				ports: [ '3306' ],
 				environment: {
-					MYSQL_ALLOW_EMPTY_PASSWORD: 'yes',
+					MYSQL_ROOT_PASSWORD: 'password',
+					MYSQL_DATABASE: 'wordpress',
```
Comment (youknowriad) [inline packages/env/lib/build-docker-compose-config.js:180]: Setting the database here, makes sure the container creates it (which was done previously in the wordpress image)

Comment (youknowriad) [reply packages/env/lib/build-docker-compose-config.js:180]: This is also why we need a separate container for the database of the tests env.

**Comment by gwwar** [comment]
Have folks tested pointing at the previous docker images to pinpoint where things changed?

**Comment by youknowriad** [comment]
I think it changed in "latest" so probably not tagged yet. I think ideally, `wp-env` shouldn't be relying on a "wordpress" image since itself downloads wordpress based on some config, it could be just based on a simpler/lighter php image but that's a discussion for another time. (I didn'try but yea we could potentially fix  the version to **wordpress:5.7** image. 

I'm going to end my day soon (and be AFK tomorrow) so I'm passing on the batton on this one.

**Comment by ribaricplusplus** [comment]
Don't know if this is useful information, just wanted to let you know that I tried the fix as it is now and it doesn't work for me. This is what I get now:

```
bruno@brunopc:[tetenv]> npx wp-env start
⚠ Warning: could not find a .wp-env.json configuration file and could not determine if '/home/bruno/Everything/tetenv' is a WordPress installation, a plugin, or a theme.
✖ Error while running docker-compose command.
Creating 6702d4ebf74b8790de999cf3da46ad4b_cli_run ... done
mysqlcheck: Got error: 1045: Access denied for user 'root'@'172.24.0.6' (using password: YES) when trying to connect
npm ERR! code 2
npm ERR! path /home/bruno/Everything/tetenv
npm ERR! command failed
npm ERR! command sh -c wp-env "start"

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/bruno/.npm/_logs/2021-03-11T16_36_02_579Z-debug.log

```
I'm using just a plain directory with `@wordpress/env` (this PR version) installed.

Some system info:

OS: Linux
Node v15.8.0
Docker 20.10.5
Docker-Compose v1.27.4

**Comment by pbking** [comment]
I tested this locally with a .wp-env.json file and that did seem to address the error.  I'm not sure of the most appropriate way to install this from a particular branch but the mechanism I used was just `npm install -g .`

Is there a better way?

**Comment by ribaricplusplus** [comment]
@pbking By 'address the error' do you mean that it worked when you used a particular `.wp-env.json` file? If so could you share that file?

**Comment by pbking** [comment]
Sure thing.  I also realized that `npm link .` is probably what have installed that correctly so I can't promise that my test was legit.  I'm retesting now to make sure it actually did what I said. :P

**Comment by pbking** [comment]
```
{
  "core": "WordPress/WordPress#5.6-branch",
  "plugins": [
    "https://github.com/WordPress/gutenberg/releases/download/v10.1.1/gutenberg.zip"
  ],
  "env": {
    "development": {
      "port": 3333
    },
    "tests": {
      "port": 3334
    }
  }
}
```

This was my .wp-env.json file I ran from [this location](https://github.com/Automattic/themes/tree/trunk/mayland-blocks).

With the above patch linked via `npm link .` as described above this seems to work as expected.
I seem to be having less success if I don't specify a version of core though.

It was likely specifying the version of CORE to use that got it to work in the first place.

**Comment by ribaricplusplus** [comment]
Interesting, I still get the same old error. Though tests seem to be getting the error that you mention.

**Comment by pbking** [comment]
So yea, I can confirm that... things still aren't working out for me.

I was ultimately able to reliably work around this situation by:

* npm linking npm link . from the gutenberg/packages/env folder (after uninstalling @wordpress/env globally first)
* ALSO specifying a CORE version in the .wp-config[.override].json file

To get my a8c-wp-env to work I couldn’t use `npm start` as that seems to use exclusively the version of @wordpress/env that is released on npm.  So from the a8c-wp-env project folder `wp-env start` instead seems to have worked to launch the environment in my ‘last configured state’ and also with the version of the plugin I linked to.

There were also plenty of wp-env destroy calls in there as well.

So I limped myself to a working(ish) dev environment.  And I'm afraid troubleshooting it further is well over my head. :/

**Comment by ockham** [comment]
> Have folks tested pointing at the previous docker images to pinpoint where things changed?

Experimenting here: https://github.com/WordPress/gutenberg/pull/29798

**Comment by ockham** [comment]
> Have folks tested pointing at the previous docker images to pinpoint where things changed?

Found what I believe is the relevant PR: https://github.com/docker-library/official-images/pull/9772
So I guess the issue is present in the 5.7 image? :thinking:

**Comment by ribaricplusplus** [comment]
#29800 might be a full fix, I'm working on a cleaner implementation before converting from draft.

**Comment by ockham** [comment]
Looking at the `wp-config.php is not writable` errors that we're seeing, [this comment](https://github.com/docker-library/wordpress/pull/557#issuecomment-756453476) seems relevant:

> For users who are changing `WORDPRESS_*` variables and expecting our `docker-entrypoint.sh` to update them, the "migration" path is to remove your `wp-config.php` file and let the new entrypoint copy in the updated version, and everything will "just work" from then on.

Not quite sure what that means for us, but I guess it's mostly relevant for these bits: https://github.com/WordPress/gutenberg/blob/e6d3f292b7c465b910869cd6506d4125108b3169/packages/env/lib/build-docker-compose-config.js#L62-L99

**Comment by ockham** [comment]
> Looking at the `wp-config.php is not writable` errors that we're seeing, [this comment](https://github.com/docker-library/wordpress/pull/557#issuecomment-756453476) seems relevant:
> 
> > For users who are changing `WORDPRESS_*` variables and expecting our `docker-entrypoint.sh` to update them, the "migration" path is to remove your `wp-config.php` file and let the new entrypoint copy in the updated version, and everything will "just work" from then on.
> 
> Not quite sure what that means for us, but I guess it's mostly relevant for these bits:
> 
> https://github.com/WordPress/gutenberg/blob/e6d3f292b7c465b910869cd6506d4125108b3169/packages/env/lib/build-docker-compose-config.js#L62-L99

Ah, I might be mistaken. The `Error: Could not process the 'wp-config.php' transformation` error seems to come from WP CLI, so this line might be a more likely candidate: https://github.com/WordPress/gutenberg/blob/e6d3f292b7c465b910869cd6506d4125108b3169/packages/env/lib/wordpress.js#L86-L88

**Comment by ockham** [comment]
> Ah, I might be mistaken. The `Error: Could not process the 'wp-config.php' transformation` error seems to come from WP CLI, so this line might be a more likely candidate:
> 
> https://github.com/WordPress/gutenberg/blob/e6d3f292b7c465b910869cd6506d4125108b3169/packages/env/lib/wordpress.js#L86-L88

So it seems like these lines: https://github.com/WordPress/gutenberg/blob/e6d3f292b7c465b910869cd6506d4125108b3169/packages/env/lib/wordpress.js#L79-L90

are used to add config settings like these to the respective `wp-config.php`s:

https://github.com/WordPress/gutenberg/blob/e6d3f292b7c465b910869cd6506d4125108b3169/packages/env/lib/config/config.js#L78-L87

It seems like the more 'idiomatic' way that the WP Docker image provides to pass extra config settings is the `WORDPRESS_CONFIG_EXTRA` var (see e.g. https://hub.docker.com/_/wordpress or https://www.mrjamiebowman.com/software-development/wordpress/docker-compose-wordpress_config_extra/).

Maybe we can use that, and drop the [logic that attempts to set these via WP CLI](https://github.com/WordPress/gutenberg/blob/e6d3f292b7c465b910869cd6506d4125108b3169/packages/env/lib/wordpress.js#L79-L90).

**Comment by ribaricplusplus** [comment]
#29800 Seems to fix the issue.

The PR fixes the problem but doesn't use `docker-compose` environment variable for setting up configuration as @ockham suggested.

**Comment by noisysocks** [comment]
> I think ideally, wp-env shouldn't be relying on a "wordpress" image since itself downloads wordpress based on some config, it could be just based on a simpler/lighter php image but that's a discussion for another time.

Could definitely look into this. We'd have to drop support for setting `"core": null` in `.wp-env.json` but that's no big deal.  IIRC the `wordpress` image properly installs all of the PHP extensions that WordPress needs which is why it's convenient to use. I'm sure we can make `php` work though.

> It seems like the more 'idiomatic' way that the WP Docker image provides to pass extra config settings is the WORDPRESS_CONFIG_EXTRA var (see e.g. https://hub.docker.com/_/wordpress or https://www.mrjamiebowman.com/software-development/wordpress/docker-compose-wordpress_config_extra/).

IIRC `WORDPRESS_CONFIG_EXTRA` only sets the config when `wp-config.php` is **first created**. This isn't good enough for `wp-env` which supports users providing their own `wp-config.php` and users modifying the config in their `.wp-env.json` and then restarting.

**Comment by noahtallen** [comment]
> We'd have to drop support for setting "core": null in .wp-env.json but that's no big deal. IIRC the wordpress image properly installs all of the PHP extensions that WordPress needs which is why it's convenient to use. I'm sure we can make php work though.

I've wondered if it would make sense to rely on a single Docker image, instead of different ones for every environment. For example, we have separate phpunit and CLI instances -- it seems like we could bundle CLI/phpunit into the other services if we used a single Docker image.

**Comment by noisysocks** [comment]
> I've wondered if it would make sense to rely on a single Docker image, instead of different ones for every environment. For example, we have separate phpunit and CLI instances -- it seems like we could bundle CLI/phpunit into the other services if we used a single Docker image.

Yes absolutely! The `wordpress` image is optimised for deploying production versions of WordPress and so doesn't include developer tools. Ideally we would have a `wordpress-develop` image based on PHP which contains WordPress, WP CLI, phpunit, xdebug, etc.

**Comment by ribaricplusplus** [comment]
Might be worth considering that Docker is meant to be used 1 process per container. If it is desired to have a particular PHP environment shared between images, that can be accomplished by creating a base image which is then extended for every process. Might not be the best idea to bundle everything into a single image.

An article that talks a bit about this: https://developers.redhat.com/blog/2016/02/24/10-things-to-avoid-in-docker-containers/

**Comment by noahtallen** [comment]
Closing in favor of #29800.

**Pull request closed without merging**